web/packages/extension/src/popup.ts

```diff
@@ -83,28 +83,48 @@ async function queryTabStatus(
     optionsChanged();
 }
 
-function objectsEqual<T>(x: NonNullable<T>, y: NonNullable<T>) {
-    for (const [key, value] of Object.entries(x)) {
-        if (y[key as keyof T] !== value) {
-            return false;
+/**
+ * Should only be called on data type objects without any "cyclic members" to avoid infinite recursion.
+ */
+function dataEqual(x: unknown, y: unknown): boolean {
+    if (
+        typeof x === "object" &&
+        typeof y === "object" &&
+        x !== null &&
+        y !== null
+    ) {
+        // Two non-null objects.
+
+        for (const [key, value] of Object.entries(x)) {
+            const value2 = y[key as keyof typeof y];
```

Comment:
For slightly more compact code, I prefer to "inline" `value2` directly into the `dataEqual()` call:
```ts
if (!dataEqual(value, y[key as keyof typeof y])) {
    return false;
}
```
(here and right below)